### ISSUE TITLE
Extract method body after guards into a separate method

### ISSUE THREAD

**Issue by hazzik** [maca88/AsyncGenerator#1]
The sync methods with guard exceptions/preconditions shall be split into two methods:

```csharp
public int Divide(int a, int b) {
   if (b==0) throw new InvalidOperationException();
   return a/b;
}
```
```csharp
public Task<int> DivideAsync(int a, int b) {
   if (b==0) throw new InvalidOperationException();
   return DivideInternalAsync(a, b);
}
// Not sure what shall be the order of "Internal" and "Async" postfixes
private async Task<int> DivideInternalAsync (int a, int b) {    
  return await Task.FromResult(a/b);
}
```

**Comment by maca88** [comment]
There is no need to have a separate method for the given example as it has no benefits, only drawbacks.
The proposed solution has few issues and drawbacks:
- Method `DivideAsync` will return a faulty `Task` in case `b` is `0` (see the example below)
- await in `DivideInternalAsync` is not needed, we can just return the Task without awaiting it, as we know that the division will never throw
- Adding the `DivideInternalAsync` method does not solve anything, it is just one more call

For easier understanding, I created a [practical example](https://dotnetfiddle.net/npiN3q) with different async implementations including yours.
To summarize, the ideal async counterpart for the given example would be:
```
public Task<int> DivideAsync(int a, int b)
{
    if (b == 0) return Task.FromException<int>(new InvalidOperationException());
    return Task.FromResult<int>(a / b);
}
```

**Comment by hazzik** [comment]
>await in DivideInternalAsync is not needed, we can just return the Task without awaiting it, as we know that the division will never throw

The task there is just for simplicity, there suppose to be a real async operation, sorry for the confusion.

**Comment by hazzik** [comment]
The main purpose of the such a method split is to throw a naked exception without wrapping it into a task. And throw as soon as the method is called, and not when the task will be awaited.

Please read the provided [article](http://www.luxoft-training.com/news/errorprone-net-part-3/).

**Comment by hazzik** [comment]
```
using System;
using System.Linq;
using System.Threading.Tasks;

using Nito.AsyncEx;


public class TestAsync1
{
	public static Task<int> DivideAsync(int a, int b)
	{
		if (b == 0) throw new InvalidOperationException();
		return DivideInternalAsync(a, b);
	}

	private static async Task<int> DivideInternalAsync(int a, int b)
	{
		await Task.Delay(42);
		return a / b;
	}
}

public static class TestAsync2
{
	public static async Task<int> DivideAsync(int a, int b)
	{
		if (b == 0) throw new InvalidOperationException();
		await Task.Delay(42);
		return a / b;
	}
}


public static class Program
{
	public static void Main()
	{
		AsyncContext.Run(() => MainAsync());
	}

	public static async void MainAsync()
	{
                //var t = TestAsync1.DivideAsync(1, 0);
		var t = TestAsync2.DivideAsync(1, 0);
		Console.WriteLine("Hello world"); // depending on what line you have uncommented you'll see this line executed, or not.
		await t;
	}
}```

**Comment by hazzik** [comment]
Here some examples from the ~gurus~ Microsoft: https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/File.cs#L753

**Comment by maca88** [comment]
Thanks for the article, now I understand the reason why. I will do my best to support it.

**Comment by hazzik** [comment]
Also, I think it is good to add an option to make these methods a local functions. It seems that local functions were proposed for exactly this reason: http://thebillwagner.com/Blog/Item/2016-03-02-C7FeatureProposalLocalFunctions

**Comment by maca88** [comment]
Agreed, I will add an option in the transformation configuration like `UseLocalFunctions` which will be false by default. Also, I need to add them into the analyzation process as currently only anonymous functions are recognized inside a method.

**Comment by hazzik** [comment]
Fixed by dd45dd8 + c97945d
